### PR TITLE
Remove `vim-repeat` dependency to fix issue #1238

### PR DESCRIPTION
**Pull request by lacygoill**
## Introduction

This PR tries to eliminate `vim-repeat` as a dependency, in order to fix the [issue 1238](https://github.com/lervag/vimtex/issues/1238).  It affects these `<plug>` key sequences used as a lhs in some mappings:

| <plug> key sequence                            | mapped by default to | installed in file         |
| -----------------------------------------------|----------------------|---------------------------|
| `<plug>(vimtex-cmd-change)`                    | `csc`                | autoload/vimtex/cmd.vim   |
| `<plug>(vimtex-cmd-delete)`                    | `dsc`                | autoload/vimtex/cmd.vim   |
| `<plug>(vimtex-cmd-create)`                    | `F7`                 | autoload/vimtex/cmd.vim   |
| `<plug>(vimtex-cmd-toggle-star)`               | `tsc`                | autoload/vimtex/cmd.vim   |
| <br>                                           |                      |                           |
| `<plug>(vimtex-delim-toggle-modifier)`         |  `tsd`               | autoload/vimtex/delim.vim |
| `<plug>(vimtex-delim-toggle-modifier-reverse)` |  `tsD`               | autoload/vimtex/delim.vim |
| `<plug>(vimtex-delim-delete)`                  |  `dsd`               | autoload/vimtex/delim.vim |
| `<plug>(vimtex-delim-change-math)`             |  `csd`               | autoload/vimtex/delim.vim |
| <br>                                           |                      |                           |
| `<plug>(vimtex-env-delete)`                    | `dse`                | autoload/vimtex/env.vim   |
| `<plug>(vimtex-env-delete-math)`               | `ds$`                | autoload/vimtex/env.vim   |
| `<plug>(vimtex-env-toggle-star)`               | `tse`                | autoload/vimtex/env.vim   |
| `<plug>(vimtex-env-change)`                    | `cse`                | autoload/vimtex/env.vim   |
| `<plug>(vimtex-env-change-math)`               | `cs$`                | autoload/vimtex/env.vim   |

---

I've tried to respect the style of the original code (indentation level, explicit `l:` scope, full name of commands, folding markers ...), but I may have missed something.

There may be too much code, and even though I made some manual tests, I'm not sure it's completely reliable, so feel free to close the PR if it can't be merged.

Some questions/answers about how the changes work in the following.

## Why executing `:normal` to press `g@l`?

IOW, why this:

```vim
nnoremap <silent><buffer> <plug>(vimtex-cmd-delete)
    \ :<c-u>call <sid>setup_operator('delete')<bar>normal! g@l<cr>
                                                   ^^^^^^^^^^^
```

Instead of this:

```vim
nnoremap <silent><buffer> <plug>(vimtex-cmd-delete)
    \ :<c-u>call <sid>setup_operator('delete')<cr>g@l
                                                  ^^^
```

Here, both versions work, but the second one won't work if `s:normal_mode()` calls `input()`. A MWE:

```vim
nno  cd  :<c-u>call Func()<cr>bbb
fu! Func() abort
    call input('>')
endfu
```

Press `cd`, and you'll see that `bbb` has been consumed by `input()` and inserted on the command-line, instead of being pressed.  But if you rewrite the mapping like this, it works as expected:

```vim
nno  cd  :<c-u>call Func()<bar>norm! bbb<cr>
fu! Func() abort
    call input('> ')
endfu
```

To stay consistent, I've always used the first version, even when it was not necessary, so that if you change the code later and add an `input()` somewhere, it won't consume `g@l`.

## Why changing `vimtex#cmd#create_ask()`?

Before, this function was called by pressing `F7` or the dot command. In both cases, it asked the user for a command name.

However, when the user pressed `.`, they didn't have to type anything on the input prompt because `repeat#set()` wrote the last command name input by the user inside the typeahead buffer, followed by a CR, and so `input()` consumed the latter automatically. As a result, the same function `create_ask()` could behave differently depending on how it was called without inspecting any variable.

I could write the command name in the typeahead buffer, but I would need to do it only when the user pressed the dot command. I don't know how to do that without remapping `.` which would lead to re-write `vim-repeat` and defeat the purpose of this PR.

So, I had to change the signature of `create_ask()` to pass it a second flag, that it can inspect to decide whether it should prompt the user for a command name. Note that not only does it fix the issue `#1238` but it also removes the brief display of the input prompt which may be distracting to some users.

## Do the commands support a count?

No they don't.

I could have added support for a count, but:

 - It would have added a lot of code/complexity.
 - In the [latest commit](https://github.com/lervag/vimtex/commit/34e2423391daf614123f618910eaaedcc60a10cd), you have prefixed some of the `:call` used in mappings with `C-u`, which kills the repetition of the called functions, and the latters don't inspect `v:count[1]`; so I assume you didn't intend these commands to support a count.
 - I've searched for the pattern `count` in `:h vimtex`, and couldn't find a match where the documentation was saying that the commands are meant to support a count.
 - Even if it's not strictly equivalent, one can still press `.` to repeat a command a few times.

## Why removing `vimtex#echo#input()` from `vimtex#cmd#change()`?

`vimtex#cmd#change()` must be called by an operator function (`s:opfunc()`) to be made dot repeatable. And, inside an opfunc, I don't know how to print a message, then prompt the user for some input, without the message being erased. The only way I found to maintain the old behavior of the command, was to move `input()` outside the opfunc.

MWE:

```vim
nno cd :set opfunc=Func<cr>g@l
fu! Func(_)
    echo 'question'
    let answer = input('>')
endfu
" Press `cd`.
" You'll only see the prompt (`>`), not the message (`question`).
```

I think the issue comes from Vim which, inside an operator function, automatically redraws before any command which prints a message (like `:echo`, or `:call input()`). I don't know how to prevent it. `:h echo-redraw` doesn't help here. Adding a `:redraw` before the call to `input()` doesn't prevent Vim from erasing the previous message.

Just as an experiment, I also tried to print 2 messages with 2 `:echo` from a timer:

```vim
" ✔
:echo 'foo' | echo 'bar'
"    → foo
"      bar

" ✘
:call timer_start(0, {-> execute('echo "foo" | echo "bar"', '')})
"    → bar
```

And it suffers from the same issue. Same thing from an autocmd:

```vim
augroup test_echo
    au!
    au CursorHold * call Func() | au! test_echo
augroup END
fu! Func()
    echo 'foo'
    echo 'bar'
endfu
" ✘ wait a few seconds, only 'bar' is printed
```

## Why removing `vimtex#env#change()` from `vimtex#env#change_prompt()`?

For the same reason as before, I need `input()` to be called from outside an opfunc. So, I call `vimtex#env#change_prompt()` from `s:setup_operator()` which is called before the opfunc, then the latter can call `vimtex#env#change()` to make the change.

## Why checking `a:0` in `vimtex#env#change()`?

Originally, `vimtex#env#change()` could be called by `vimtex#env#toggle_star()` and be passed 3 arguments. It's still the case. However, it could also be called by `vimtex#env#change_prompt()`. It's not the case anymore, because again, `input()` must be called from outside an opfunc. So, `s:setup_operator()` calls `vimtex#env#change_prompt()` to get the user's input, then later `vimtex#env#change()` is called to make the change. However, I can't pass the 3 arguments that `vimtex#env#change()` expected originally. I could pass the user input (`a:new`), but not the other 2 arguments (`a:open`, `a:close`), because, contrary to the input which can be re-used when `.` is pressed, those must be recomputed for every repeat.

To summarize, when `vimtex#env#change()` is called by `vimtex#env#toggle_star()`, it can be passed its original 3 arguments, but not when it's called by `vimtex#env#change_prompt()`. That's why the function must check `a:0` to determine from where it was called and get its arguments accordingly.

### CODE REVIEW

**Comment by andymass** [comment]
> but the second one won't work if s:normal_mode() calls input().

Would `inputsave()`/`inputrestore()` help here?  (I haven't thought too hard about it though).

> I've searched for the pattern count in :h vimtex

The `tsd`/`tsD` are supposed to accept a count:

https://github.com/lervag/vimtex/blob/34e2423391daf614123f618910eaaedcc60a10cd/doc/vimtex.txt#L2364

I wonder, since these particular maps do not take user input, would adding the count back be possible?  Maybe,

       \ :<c-u>call <sid>setup_operator('...')<bar>execute 'normal!' (v:count ? v:count : '') . 'g@l'<cr>

or 

       \ :<c-u>call <sid>setup_operator('...')<bar>normal! <c-r>=v:count ? v:count : ''<cr>g@l<cr>

**Comment by lacygoill** [comment]
@andymass Thank you very much for the comments.

> Would inputsave()/inputrestore() help here? (I haven't thought too hard about it though).

It would help, but I tried to minimize the number of lines of code.
Using `:normal! g@l` should be more future-proof without adding 2 lines of code in each opfunc.

> The tsd/tsD are supposed to accept a count:

Ah yes you're right, I missed this part or didn't see the link between it and `tsd`/`tsD`.

> I wonder, since these particular maps do not take user input, would adding the count back be possible? Maybe,
>
>    `\ :<c-u>call <sid>setup_operator('...')<bar>execute 'normal!' (v:count ? v:count : '') . 'g@l'<cr>`
>
> or
>
>    `\ :<c-u>call <sid>setup_operator('...')<bar>normal! <c-r>=v:count ? v:count : ''<cr>g@l<cr>`

I think they should both work, I have commited a second commit using the second version, as it's shorter, and squashed the 2 commits.

**Comment by lervag** [comment]
@lacygoill Thank you for this, I think you have done a pretty good job here. I took the liberty of doing some minor modification of your formatting, as it made it easier for me to read and understand what you've written. I hope you don't mind. For future updates, please note that you should keep a paragraph without line breaks for optimal text flow on github.

After the latest commit and input from @andymass, I can't see anything wrong with your arguments. But the tests are failing. I investigated some, and I've attached a patch that should fix the failing tests (copy the diff to a file, then use `git apply`). Finally, I will also need to do a review of the code before merging.

If I understand correctly, you are claiming that this PR should make vimtex completely independent of vim-repeat, right?

```diff
diff --git a/test/vader/cmd.vader b/test/vader/cmd.vader
index 1797a85..62e88ef 100644
--- a/test/vader/cmd.vader
+++ b/test/vader/cmd.vader
@@ -11,8 +11,8 @@ Expect tex (Cmd: verify):
   foobar
   foo
 
-Do (Cmd: dsc):
-  dsc
+Execute (Cmd: dsc):
+  execute "normal dsc"
 
 Expect tex (Cmd: verify):
   foo
diff --git a/test/vader/delims.vader b/test/vader/delims.vader
index 080c2e5..1c0f581 100644
--- a/test/vader/delims.vader
+++ b/test/vader/delims.vader
@@ -4,8 +4,8 @@ Given tex (Delims: toggle in complicated math):
          & 1 \\
   \end{smallmatrix}\bigr)$
 
-Do (Toggle):
-  3jtsd
+Execute (Toggle):
+  execute 'normal 3jtsd'
 
 Expect tex (Verify):
   $(\begin{smallmatrix}
diff --git a/test/vader/env.vader b/test/vader/env.vader
index 823ed95..d7dfdc3 100644
--- a/test/vader/env.vader
+++ b/test/vader/env.vader
@@ -15,8 +15,8 @@ Given tex (Environments: foo and bar):
     Bar
   \end{bar}
 
-Do (Change env: foo -> baz):
-  csebaz\<cr>
+Execute (Change env: foo -> baz):
+  execute "normal csebaz\<cr>"
 
 Expect tex (Verify):
   \begin{baz}
@@ -27,8 +27,8 @@ Expect tex (Verify):
     Bar
   \end{bar}
 
-Do (Change env: foo -> bar (with completion)):
-  cse\<c-z>\<c-z>\<c-z>\<cr>
+Execute (Change env: foo -> bar (with completion)):
+  execute "normal cse\<c-z>\<c-z>\<c-z>\<cr>"
 
 Expect tex (Verify):
   \begin{bar}
@@ -47,8 +47,8 @@ Given tex (Environments: test and center):
     \begin{center} a \end{center}
   \end{test}
 
-Do (Delete env: test and center):
-  dsedse
+Execute (Delete env: test and center):
+  execute 'normal dsedse'
 
 Expect tex (Verify):
      a 
```

File: autoload/vimtex/cmd.vim [hunk 1]
```diff
@@ -231,9 +221,38 @@ function! vimtex#cmd#toggle_star() " {{{1
...
-  " Restore cursor position and create repeat hook
+  " Restore cursor position
   cal vimtex#pos#set_cursor(l:save_pos)
-  silent! call repeat#set("\<plug>(vimtex-cmd-toggle-star)", v:count)
+endfunction
+
+" }}}1
+
+function! s:setup_operator(operator) abort " {{{1
+  let &opfunc = s:snr().'opfunc'
```
Comment (lervag) [inline autoload/vimtex/cmd.vim]: 1. I prefer to keep space before/after `.` (makes it more distinguishable from dictionary dots).

2. I assume that `s:snr()` is the standard way of accessing script local functions when we've lost the scope. Just curious, are you sure we need it, i.e., that we can't just do `s:opfunc`?

File: autoload/vimtex/cmd.vim [hunk 2]
```diff
@@ -231,9 +221,38 @@ function! vimtex#cmd#toggle_star() " {{{1
...
+  let s:operator = a:operator
+  if s:operator isnot# 'change' && s:operator isnot# 'create_ask' | return | endif
+  if s:operator is# 'change' && empty(vimtex#cmd#get_current()) | return | endif
+  let s:cmd_name_{a:operator} = vimtex#echo#input({
+        \ 'info' : ['Change command: ',
+        \   ['VimtexWarning',
+        \     a:operator is# 'change' ? vimtex#cmd#get_current().name : '(empty to cancel)']],
+        \})
+  let s:cmd_name_{a:operator} = substitute(s:cmd_name_{a:operator}, '^\\', '', '')
+endfunction
```
Comment (lervag) [inline autoload/vimtex/cmd.vim]: One idea for minor improvement is to make `s:setup_operator` and `s:opfunc` methods of a "class" `s:operator`. Call it semi-OO, if you want. This way, there would be less `s:` variables, since `s:setup_operator` could save state to itself. Something like this:

```vim
let s:operator = {}

function! s:operator.setup(operator) dict abort
  "...
endfunction

function! s:operator.opfunc(_) dict abort
￼ execute 'call vimtex#cmd#' . {
￼       \   'change': 'change(get(s:, "cmd_name_change", ""))',
￼       \   'create_ask': 'create_ask(0)',
￼       \   'delete': 'delete()',
￼       \   'toggle_star': 'toggle_star()',
￼       \ }[self.operator]
endfunction
```

Another benefit of this is that it would make it slightly easier to read the code when e.g. `vimtex#cmd#change` uses a script local variable, since it would use `s:operator.<var>`, which would instantly be read as something that is coupled to the operator code.

File: autoload/vimtex/cmd.vim [hunk 3]
```diff
@@ -231,9 +221,38 @@ function! vimtex#cmd#toggle_star() " {{{1
...
   cal vimtex#pos#set_cursor(l:save_pos)
-  silent! call repeat#set("\<plug>(vimtex-cmd-toggle-star)", v:count)
+endfunction
+
+" }}}1
+
+function! s:setup_operator(operator) abort " {{{1
+  let &opfunc = s:snr().'opfunc'
+  let s:operator = a:operator
+  if s:operator isnot# 'change' && s:operator isnot# 'create_ask' | return | endif
```
Comment (lervag) [inline autoload/vimtex/cmd.vim]: Is there a good reason to use `is` and `isnot` instead of `==` and `!=`?

File: autoload/vimtex/delim.vim [hunk 4]
```diff
@@ -202,6 +199,46 @@ endfunction
...
+          \   ['Change surrounding delimiter: ', ['VimtexWarning', l:name]],
+          \ 'complete' : 'customlist,vimtex#delim#change_input_complete',
+          \})
+  endif
+endfunction
+
+" }}}1
+function! s:opfunc(_) abort " {{{1
+  if s:operator is# 'toggle_modifier'
+    call call('vimtex#delim#toggle_modifier', [s:toggle_modifier_dir])
```
Comment (lervag) [inline autoload/vimtex/delim.vim]: Can you not just call this directly?

File: autoload/vimtex/delim.vim [hunk 5]
```diff
@@ -202,6 +199,46 @@ endfunction
...
+  else
+  execute 'call vimtex#delim#'
```
Comment (lervag) [inline autoload/vimtex/delim.vim]: Wrong indentation.

**Comment by lervag** [comment]
Great, thank you for the updates!

**Comment by lervag** [comment]
Thank you for the useful input! I'll look into the possibility of making this slightly more maintable, but it seems that the gain might be lost in the extra burden.

Also, I was not aware of the distinction between making the dictionary functions directly and between defining them and allocating them wrt. the debugging part. This is useful, and I agree it makes a lot of sense to avoid defining functions that are not named when debugging.

**Comment by lervag** [comment]
Again, thank you! I really appreciate your contribution!

**Comment by lervag** [comment]
I've attempted to make some modifications that made sense to me. I mostly kept the structure you suggested, but I tried to avoid using script local variables as much as possible, keeping them inside the operator wrapper functions.
